### DISCUSSION TITLE
How to specify search space

### DISCUSSION THREAD

**Discussion by hanhanwu** [microsoft/FLAML#159] [Q&A] [answered]
Hello FLAML team,

I have a few questions about self-specified estimator's params and self-specified search space:

1. Is there anyway to specify the params of an estimator by myself? If so, how to do that?
2. Is there anyway to specify the search space of an estimator, so that I can specify a list of values for each parameter? If so, how to do that?
3. When there are multiple estimators, how can we specify the search space each each estimator?

Thank you very much!

**Comment by qingyun-wu** [comment, marked as answer]
Hi @hanhanwu,

You can specify the parameters and also the hyperparameter search space of an estimator by adding a customized learner. If there are multiple estimators, you can just specify the search space for each one of them and add each of the estimators to the `estimator_list` after calling the `add_learner(..)` function of the AutoML instance. 

Please find guidance on how to do them in this tutorial: [FLAML with different levels of customization](https://www.youtube.com/watch?v=bJfDJhe-O-c). 
You can find a code example in the following notebooks: https://github.com/microsoft/FLAML/blob/main/notebook/flaml_automl.ipynb (Section 3. Customized Learner).

**Reply by hanhanwu** [reply]
Hi @qingyun-wu ,

I just tried to specify the search space of my own LGBM, by adding some params dind't appear in FLAML's LGBM's search space:
![code1](https://user-images.githubusercontent.com/4024769/128976435-e3c504c9-9557-477f-b5aa-52d1520a559b.PNG)

But do you know why the output estimator doesn't show the params I have added?
![output](https://user-images.githubusercontent.com/4024769/128976473-cc9be6ba-2fd4-4671-9c1c-bc7e674c72b9.PNG)

**Reply by qingyun-wu** [reply]
Hi @hanhanwu,

There are two things you need to fix:
1. After creating a customized leaner, you should call `add_leaner` to add your leaner into flaml's consideration before `fit()`. In your case, add 
`automl30.add_learner(learner_name='my_lgbm', learner_class=my_lgbm)` 

before `automl30.fit(..)`.
2. The way you create the learner is not legitimate to be used in flaml: the new class should inherit the abstract class `BaseEstimator` from flaml instead of the LGBMClassifier from lightgbm. 

Please let me know if you still encounter problems after fixing these issues.

Thanks!

**Reply by hanhanwu** [reply]
I see, now the problems solved. Thank you @qingyun-wu !
